libp2p/transport/udp/udp.py

```diff
@@ -0,0 +1,163 @@
+import asyncio
+from typing import Any, Dict, List, Tuple
+
+from multiaddr import Multiaddr
+
+import asyncio_dgram
+from asyncio_dgram.aio import DatagramStream
+from libp2p.network.connection.raw_connection import RawConnection
+from libp2p.network.connection.raw_connection_interface import IRawConnection
+from libp2p.transport.exceptions import OpenConnectionError
+from libp2p.transport.listener_interface import IListener
+from libp2p.transport.stream_interface import IStreamReader, IStreamWriter
+from libp2p.transport.transport_interface import ITransport
+from libp2p.transport.typing import THandler
+from libp2p.transport.udp.udp_stream import (
+    UDPServerStreamReader,
+    UDPStreamReader,
+    UDPStreamWriter,
+)
+
+
+async def open_datagram_connection(
+    host: str, port: int
+) -> Tuple[IStreamReader, IStreamWriter]:
+    stream = await asyncio_dgram.connect((host, port))
+    reader, writer = UDPStreamReader(stream), UDPStreamWriter(stream)
+    return reader, writer
+
+
+def start_udp_server(handler: THandler, host: str, port: int) -> "UDPServer":
+    server = UDPServer(handler, host, port)
+    asyncio.create_task(server.listen())
+    return server
+
+
+class UDPServer:
+    _host: str
+    _port: int
+    _stream: DatagramStream
+    _handler_queues: Dict[Any, asyncio.Queue[bytes]]
+    _running_handlers: List[asyncio.Task[Any]]
+
+    def __init__(self, handler: THandler, host: str, port: int):
+        _handler: THandler  # in class-level, MyPy raises error, because it treats like a method
+        self._handler = handler
+        self._host = host
+        self._port = port
+        self._handler_queues = {}
+        self._running_handlers = []
+
+    async def listen(self) -> None:
+        self._stream = await asyncio_dgram.bind((self._host, self._port))
+        while True:
+            data, addr = await self._stream.recv()
+            await self._deliver_data(addr, data)
+
+    async def close(self) -> None:
+        self._stream.close()
+        # TODO: Correct way to destroy all handlers?
+        for task in self._running_handlers:
+            task.cancel()
```

Comment:
The *proper* pattern here according to the asyncio docs is:

```python
task.cancel()
try:
    await task
except asyncio.CancelledError:
    pass
```

This gives the loop the opportunity to inject the cancellation exception into the task so that it doesn't end up being reported as an unretrieved exception.

Reply:
Thanks! I missed it. And probably there are more things, I made without much revising.